git-ftp

```diff
@@ -1831,7 +1831,7 @@ do
 		*)
 			# Pass thru anything that may be meant for fetch.
 			if [ -n "$1" ]; then
-				if [ -z "$URL" ]; then
+				if [ -z "$URL" ] && echo "$1" | egrep -q -v '^(-[A-Za-z])|^(--[A-Za-z0-9-]+)'; then
```

Comment:
Do we need to test for the end of the option as well? I think it's matching too much and should be this:
 `egrep -q -v '^(-[A-Za-z])$|^(--[A-Za-z0-9-]+)$'`

Reply:
Sorry. What would be the difference with the `$`. I know it asserts the end of line, but how would this take effect? Isn’t `$1` always just one word? `$2` is the value, isn’t it?

Reply:
The regex `^-[A-Za-z]` matches everything that starts with `-` and a character. But it doesn't restrict anything afterwards. So it matches `-h` but also `-hello` and `-A-B-C8_*`. The `$` marks the end of the match. So `^-[A-Za-z]$` only matches strings that start with a `-` and then have one letter afterwards. And after that letter the string must be over, nothing else can come afterwards.

Reply:
Sorry, I skipped the `$` for the first pattern. This makes sense.

But what if somebody tries to do something like `-vv` (which is a valid option) but accidentally hits the key beside v and types `-bb` or he hits the button 3 times and types `-vvv`. With your pattern this would be set as `URL`, wouldn’t it? And this is exactly, what we wanted to avoid.

As described in https://github.com/git-ftp/git-ftp/issues/522#issuecomment-504765313 URL hostnames or domains can’t start with a `-`, so we would be fine with this. 
We only get problems if the user gives a url with the pattern `username:[password]@host` **AND** the username (not the password) starts with `-` or `--`, but I think in the real world this scenario would be really rare, wouldn’t it? It's at least rarer than `-vvv`.

Reply:
I agree that it would be really rare. But those are the cases that can take a long time to debug. I forgot that -vvv is a valid option. In that case I would rewrite the command like this:

```sh
egrep -q -v '^(-[A-Za-z]+)$|^(--[A-Za-z0-9-]+)$'
```

And in the case that a username starts with `-`, people will get a nice error message that `-username@example` is not a valid option and then they need to use the full URL with protocol: `ftp://-username@example`. That should work, shouldn't it?

Reply:
> people will get a nice error message that -username@example is not a valid option

In fact, no. I thought that too, but I just noticed while trying it out that there are other characters in the URL than `A-Za-z`. And therefore the pattern does not match. So we are fine here.

But I agree. Your pattern with `+` and `$` would be the best of all. But now the two OR parts are very similar so we could simplify it and make the second dash optional and use
```
egrep -q -v '^(--?[A-Za-z0-9-]+)$'
```
This would allow numbers in case of one dash, but I don’t thinks that’s a problem. What do you think?

Reply:
Oh. Have you ever tried to use a URL that follows the pattern `username:[password]@host` and is not just host?

I tried a few things and run into the following:
```
$ git ftp init -username@example
fatal:  Can't access remote 'ftp://-username'. Network down? Wrong URL? exiting...
```
The given URL is cut at the `@` character.

The line where this happens is: https://github.com/git-ftp/git-ftp/blob/ae1a6836f3a5bb674192fefd7e2cb8387cf55ab8/git-ftp#L618

The manual for `expr` says:
> If the match succeeds and the pattern contains at least one regular expression subexpression \`\`\\(...\\)'', the string corresponding to \`\`\1'' is returned;

Since `-username` matches the group only this part is returned and the rest gets lost. **So it looks like this kind of compact URLs is not supported at all at the moment, is it?**
If this is not supported I think we should create a new enhancement issue for this.